File: python/stack/daily-temperatures.py
class Solution:
    def dailyTemperatures(self, temperatures: List[int]) -> List[int]:
        nearest = {}
        answer = []
        for i in range(30, 101):
            nearest[i] = -1

        reverse = temperatures[::-1]
        for i in range(len(reverse)):
            neighbor = float('-inf')
            for j in range(reverse[i] + 1, 101):
                if nearest[j] != -1:
                    neighbor = max(neighbor, nearest[j])

            if neighbor > float('-inf'):
                answer.append(i - neighbor)
            else:
                answer.append(0)

            nearest[reverse[i]] = i

        return answer[::-1]

File: c++/stack/daily-temperatures.cc
class Solution {
public:
    vector<int> dailyTemperatures(vector<int>& temperatures) {
        std::unordered_map<int, int> indexes;

        std::vector<int> stack = {std::numeric_limits<int>::max()};
        std::vector<int> answer;
        for (int i = temperatures.size() - 1; i > -1; --i) {
            while (stack.back() <= temperatures[i]) {
                stack.pop_back();
            }

            int day = stack.back() != std::numeric_limits<int>::max() ? indexes[stack.back()] : i;

            stack.push_back(temperatures[i]);
            indexes[temperatures[i]] = i;

            answer.push_back(day - i);
        }
        std::reverse(answer.begin(), answer.end());
        return answer;
    }
};
